### PR TITLE
Add IBM cost analysis

### PR DESCRIPTION
**Pull request by glen-84**
Summary of the changes (Less than 80 chars)

- Adds IBM cost analysis.

---

❓ Questions

- What am I doing wrong with the error handling in `CostAnalysisMiddleware`? It doesn't end the request. It's probably something silly. [@michaelstaib]

- Is it important that a fragment definition is visited after it is spread? If you spread it twice, it is only visited once.
    - It seems logical that it would be visited from the root, not after it is first spread. This makes the path generation awkward, since you end up with paths like: `query.examples.example1Field2[0].~fragment1~~fragment1.@example` by default instead of just `~~fragment1.@example` (the fragment definition starts at the root). Would changing this break things?
- For unnamed operations, I've used `o.Operation.ToString().ToLowerInvariant()` for the path element – is that okay? ✅ 
    - f.e. `query` for an unnamed query, etc.
- For fragment spreads with no `TypeCondition`, I've used `context.OutputFields.Peek().Type.TypeName()` – is that okay?
    - i.e. The name of the type of the field on which the spread is defined.
- How do we avoid requiring the developer to add a `@listSize` directive to every single list in their project? Could we somehow attach this directive to connection types by default? [@michaelstaib]
    - `RequireOneSlicingArgument` would have to be set to `false`, otherwise the user would be forced to specify `first` or `last`.
- Should we support `introspectionOnly`? (https://ibm.github.io/graphql-specs/cost-spec.html#sec-introspectionOnly) [@michaelstaib]
- Should we support `NoSchemaIntrospectionCustomRule`? (https://ibm.github.io/graphql-specs/cost-spec.html#sec-NoSchemaIntrospectionCustomRule) - yes
- For the `CostMetrics` cache, how much memory can we use, what should the default cache size be, etc.? - we deal with that in a later pr
- I used the `NonBacktracking` option for the introspection regexes to reduce the chance of performance issues – should I keep this, or allow backtracking? - keep it!
- Is there a better way to handle optional value-typed attribute arguments in `ListSizeAttribute` (for `AssumedSize`)? Currently using a nullable backing field. - looks good
- Should we sort the fieldCostByLocation / typeCostByLocation by key (path)? - lets skip it and check if people have an issue with it
- I've not yet done anything with `@skip`/`@include`. I wonder if supporting static true/false is worth it? lets revisit this for the december release

✅ To-do:

- Directives on directive arguments (blocked by #7130)
- Schema validation (blocked by [[internal](https://github.com/orgs/ChilliCream/projects/9/views/1?pane=issue&itemId=63235828)])
- ~~Support scalars (blocked by #1380)~~
- Fix paths for fragment definitions (see question above).
- ExecutionDiagnosticEventListener methods (blocked by [[internal](https://github.com/orgs/ChilliCream/projects/9/views/1?pane=issue&itemId=65506418)])

💡 Ideas

- Should we exit early as the costs exceed the maximums?
  - Under which conditions? (only in production, etc.)
- Should we log a warning when the default list size of 1 is used? (i.e. not configured with `@listSize`)

🛠️ FIXMEs: 6.

(search for `FIXME` in `src/HotChocolate/CostAnalysis/`)

### CODE REVIEW

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/Attributes/ListSizeAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        get => _assumedSize ?? 0;
+        init => _assumedSize = value;
+    }
+
+    /// <summary>
+    /// The arguments of this field with numeric type that are slicing arguments. Their value
+    /// determines the size of the returned list.
+    /// </summary>
+    public string[]? SlicingArguments { get; set; }
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/Attributes/ListSizeAttribute.cs:30]: Why are the not init?

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/CostAnalysisMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                }
+
+                if (costMetrics.FieldCost > options.MaxFieldCost)
+                {
+                    // FIXME: This is not ending the request.
+                    context.Result = ErrorHelper.MaxFieldCostReached(
+                        costMetrics.FieldCost,
+                        options.MaxFieldCost);
+
+                    return;
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/CostAnalysisMiddleware.cs:49]: This looks good with return.

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/CostAnalysisMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                if (!cache.TryGetCostMetrics(context.OperationId, out var costMetrics))
+                {
+                    PrepareContext(context, document, validatorContext);
+
+                    costAnalysisVisitor.Visit(operationDefinition, validatorContext);
+
+                    costMetrics = (CostMetrics)validatorContext.ContextData[RequestCosts]!;
+
+                    cache.TryAddCostMetrics(context.OperationId, costMetrics);
+                }
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/CostAnalysisMiddleware.cs:40]: @PascalSenn how shall we report the costs to metrics? Just like everything else with the diagnostic events?

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/Attributes/ListSizeAttribute.cs (hunk 1 above)
Comment (glen-84) [reply src/HotChocolate/CostAnalysis/src/CostAnalysis/Attributes/ListSizeAttribute.cs]: Oversight. Fixed.

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/Attributes/CostAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+namespace HotChocolate.CostAnalysis.Attributes;
+
+[AttributeUsage(
+    AttributeTargets.Class
+    | AttributeTargets.Enum
+    | AttributeTargets.Method
+    | AttributeTargets.Parameter
+    | AttributeTargets.Property
+    | AttributeTargets.Struct)]
+public sealed class CostAttribute : DescriptorAttribute
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/Attributes/CostAttribute.cs:15]: A summary here on the class would be great.

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/CostAnalysisOptions.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+namespace HotChocolate.CostAnalysis;
+
+public sealed class CostAnalysisOptions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/CostAnalysisOptions.cs:3]: This needs a summary

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostArgumentDescriptorExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.Types;
+
+namespace HotChocolate.CostAnalysis.DescriptorExtensions;
+
+public static class CostArgumentDescriptorExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostArgumentDescriptorExtensions.cs:6]: needs xml docs

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostEnumTypeDescriptorExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.Types;
+
+namespace HotChocolate.CostAnalysis.DescriptorExtensions;
+
+public static class CostEnumTypeDescriptorExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostEnumTypeDescriptorExtensions.cs:6]: needs xml docs

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostInputFieldDescriptorExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.Types;
+
+namespace HotChocolate.CostAnalysis.DescriptorExtensions;
+
+public static class CostInputFieldDescriptorExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostInputFieldDescriptorExtensions.cs:6]: needs xml docs

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostObjectFieldDescriptorExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.Types;
+
+namespace HotChocolate.CostAnalysis.DescriptorExtensions;
+
+public static class CostObjectFieldDescriptorExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostObjectFieldDescriptorExtensions.cs:6]: needs xml docs

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostObjectTypeDescriptorExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.Types;
+
+namespace HotChocolate.CostAnalysis.DescriptorExtensions;
+
+public static class CostObjectTypeDescriptorExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostObjectTypeDescriptorExtensions.cs:6]: needs xml docs

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostScalarTypeDescriptorExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.Types;
+
+namespace HotChocolate.CostAnalysis.DescriptorExtensions;
+
+public static class CostScalarTypeDescriptorExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/CostScalarTypeDescriptorExtensions.cs:6]: needs xml docs

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/ListSizeObjectFieldDescriptorExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.Types;
+
+namespace HotChocolate.CostAnalysis.DescriptorExtensions;
+
+public static class ListSizeObjectFieldDescriptorExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/DescriptorExtensions/ListSizeObjectFieldDescriptorExtensions.cs:6]: needs xml docs

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/Properties/InternalsVisibleTo.cs [hunk 13]
```diff
@@ -0,0 +1,3 @@
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("HotChocolate.CostAnalysis.Tests")]
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/Properties/InternalsVisibleTo.cs:3]: Move to csproj file

File: src/HotChocolate/CostAnalysis/src/CostAnalysis/RequestExecutorBuilderExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,91 @@
+using HotChocolate.CostAnalysis.Caching;
+using HotChocolate.CostAnalysis.Directives;
+using HotChocolate.CostAnalysis.Types;
+using HotChocolate.Execution.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace HotChocolate.CostAnalysis;
+
+public static class RequestExecutorBuilderExtensions
```
Comment (michaelstaib) [inline src/HotChocolate/CostAnalysis/src/CostAnalysis/RequestExecutorBuilderExtensions.cs:9]: needs xml docs

**Pull request merged**